torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -2954,7 +2956,12 @@ struct to_ir {
     // Push the source range of a call in case compiling this function
     // triggers an error
     ErrorReport::CallStack::update_pending_range(tree.range());
-    return emitSugaredExpr(tree, 1, type_hint)->asValue(tree.range(), method);
+    Value* out_val =
+        emitSugaredExpr(tree, 1, type_hint)->asValue(tree.range(), method);
+    if (type_hint == AnyType::get() && out_val->type() != AnyType::get()) {
```

Comment:
Do all the tests work if you remove this? If yes, could you remove this ? I 

If no, could you rewrite this to add an unchecked cast if `*out_val->type() != *type_hint && out_val->isSubtypeOf(type_hint)` ?

Reply:
It doesn't work if I remove it, since `emitSugaredExpr` will ignore the `type_hint` for the case `TK_VAR `.
And, I don't think it should be changed to `subtype` for the output. (it will raise error for my test case `M3`)
```python
class M3(torch.nn.Module):
    def forward(self) -> Optional[int]:
        return None if self.training else 1
```
It's not reasonable to cast the output value from `int` to `Optional[int]` or `None` to `Optional[int]`.
The `unchecked_cast` should be used for `Any`.

Reply:
I will verify it using 
-> Union[int, str]
1 if self.training else 'xxx'

Reply:
Interesting, `Union` is not supported in JIT. 
So, IMO we should not to cast for all subtype. In JIT, reasonable subtypes only happen for `Optional` (no cast) and `Any` (required cast)

Reply:
So, the reason we need to add the cast is because by default we do not unify two types to Any.
```
if cond:
    a = 1.
else 
    a = 1
```
I think this is reasonable - Any is hard  to use, and in most cases you want to  fix the code there so  that both branches are floats. There's actually one other case where we have a type which is a SuperType which we don't by default unify to - [NumberType](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/core/type.cpp#L249). This is an internal type though, I don't think anyone uses it. The `*out_val->type() != *type_hint && out_val->isSubtypeOf(type_hint)` is generically valid, but it's true that it only applies to `Any` right now. 

Could you add the following comment?
> AnyType is the only user-exposed type which we don't unify to from its subtypes, so we add a cast for use cases like `x : Any = 1 if cond else "str"`

Reply:
Thanks @eellison . Will add the comment and ship.